Код в соответствии с ТЗ №6

maximmedencev opened the pull request.

maximmedencev committed:
Дополнить проект в соотвествии с ТЗ №6

maximmedencev committed:
Исправить code style

maximmedencev committed:
Исправить code style

maximmedencev committed:
Исправить code style(параметры с маленькой буквы)

slavrikk commented on src/ru/yandex/practicum/tasktracker/InMemoryHistoryManager.java:
```diff
@@ -1,21 +1,87 @@
 public class InMemoryHistoryManager implements HistoryManager {
+    private Node head;
+    private Node tail;
+    private int size = 0;
```
Для чего считается size, если он нигде не используется?) Лишние операции

maximmedencev replied:
Убрал size из полей

slavrikk commented on src/ru/yandex/practicum/tasktracker/InMemoryHistoryManager.java:
```diff
@@ -1,21 +1,87 @@
+
+    private void removeNode(Node node) {
+
+        if (node == null || size == 0)
```
Если нужен проверка на размер, то можно взять из коллекции `history.size()`

maximmedencev replied:
Переделал. Теперь используется .. || history.isEmpty()

slavrikk commented on src/ru/yandex/practicum/tasktracker/InMemoryHistoryManager.java:
```diff
@@ -1,21 +1,87 @@
+        if (node == head) {
+            head = head.next;
+            head.prev = null;
+            node = null;
```
Можно не приравнивать node к null, он и так удалится из памяти

maximmedencev replied:
Убрал приравнивание к null

slavrikk commented on src/ru/yandex/practicum/tasktracker/InMemoryHistoryManager.java:
```diff
@@ -1,21 +1,87 @@
+        if (node == null || size == 0)
+            return;
+
+        if (size == 1) {
```
Думаю лишняя проверка на единичный размер, все и так должно работать

maximmedencev replied:
Исправил. Взял за основу реализации этого метода из java.util.LinkedList, но провел замеры и выяснил, что мой код показывает лучшее быстродействие, хотя и не всегда. Вот результаты замеров при разных условиях:

**Старая реализация removeNode(Node)** 
Прошло времени, нс: 3187
Прошло времени, нс: 1902
Прошло времени, нс: 1678
Прошло времени, нс: 1437

**Новая реализация removeNode(Node)** 
Прошло времени, нс: 3323
Прошло времени, нс: 1709
Прошло времени, нс: 1830
Прошло времени, нс: 1553

slavrikk commented on src/ru/yandex/practicum/tasktracker/Node.java:
```diff
@@ -0,0 +1,13 @@
+package ru.yandex.practicum.tasktracker;
+
+public class Node {
+    public Task data;
```
Все поля лучше обозначить как `private`

maximmedencev replied:
сделал поля private. Создал геттеры и сеттеры

slavrikk commented on test/ru/yandex/practicum/tasktracker/InMemoryHistoryManagerTest.java:
```diff
@@ -0,0 +1,150 @@
+    public static TaskManager inMemoryTaskManager;
+
+    public static boolean taskFieldsEquals(Task task1, Task task2){
+        if(task1.getId()==task2.getId())
```
Все условия лучше сделать в одном `if` и через `&&`

maximmedencev replied:
исправлено

maximmedencev committed:
Убрать проверку на еденичный размер в removeNode(Node)
Сделать поля класса Node приватными
Убрать поле size из InMemoryHistoryManager
Убрать приравнивание к null удаляемых объектов в InMemoryHistoryManager
